design.md

```diff
@@ -0,0 +1,128 @@
+  - We can ship the default simulator plugin with the IntelliJ plugin, but there must be a way to download any of them because of the [usability requirement](#usability).
+
+#### UI Interaction
+
+- ! RFC: The kernel must be able to notify the UI of multiple tasks happening in parallel. For example, running a script could cause some repos to get cloned and a toolchain to get downloaded in parallel. As a user, I want to see both of these operations as tasks in IntelliJ happening in parallel. As a general rule, a script can start multiple tasks.
+- ! RFC: The kernel needs a way to ask the UI for confirmation with a custom confirmation message. For example, if the kernel needs to flash the device with a new program, by default it should ask the user for confirmation. There also needs to be some local environment configuration option that bypasses this so that the confirmation results is either always allowed or denied (for headless operation). For example, you may want to always allow the kernel to flash the device during development or during some competitions (as a user, I would rather have a delayed start to my match rather than having the robot not work at all; it depends on the competition). As another example, you may want to always stop the kernel from flashing the device if you have released a product and you expect all the products to work the same for all your users and would rather handle these issues with your own support system.
+
+### Headless Execution
+
+- The kernel must not include any references to JavaFX so that it can run in a headless environment.
+
+### Data Logging
+
+- Every connected device resource must be automatically available as a data source.
+  - Sensors return their sensor value (e.g., potentiometer -> voltage, encoder -> ticks, IMU -> values of axes).
+  - Actuators return their control signal (e.g., servo -> angle, solenoid -> digital state, motor -> voltage). These are the values the user's program commands, not the measured value (which would be a sensor).
+- A script may add its own data sources via *variables*.
+  - A variable is addressed by name (a String).
+  - State changes of variables are tracked such that the current state of a variable can be read from another thread in parallel.
+- Data types that can be logged:
+  - Numerical types (e.g., int, long, float, double, byte, char, short, boolean).
+  - Strings.
+  - Frame transforms.
+  - Images (e.g., frames from a camera feed).
+  - Arbitrary bytes.
+- Data that has been logged must be able to be saved and exported to other formats (e.g., CSV) if the data type supports it (e.g., when exporting to a CSV you can export numerical data but not images).
+- The amount of data to buffer must be configurable via the settings.
+
+### Hardware
+
+- A robot can be loaded from its config file and either a connection method (if connecting to the hardware) or simulator plugin (if connecting to a simulator).
+  - ! RFC: Schema of a robot config file.
```

Comment:
Let's figure out what a robot config file needs. @madhephaestus Maybe you and I can figure this one out? Not that others can't contribute, but we have the experience here.

Reply:
Idea (by no means saying we have to go with it), but if the robot config file is JSON, you can write the schema as protobuf and then be able to serialize/deserialize it quite easily.